### DISCUSSION TITLE
Is there a way to have dynamic RETURN_TYPES ?

### DISCUSSION THREAD

**Discussion by devingfx** [Comfy-Org/ComfyUI#4800] [Q&A]
I need a node that "casts" the output depending on the inputs (essentially a format COMBO to choose the output type).
Is this possible?

**Comment by ltdrdata** [comment]
There's only an ugly wildcard trick to do that.

backend:
https://github.com/ltdrdata/ComfyUI-Impact-Pack/blob/940ad41b1508cf9156bea8ee0ef7c87a851cc24f/modules/impact/logics.py#L642

frontend:
https://github.com/ltdrdata/ComfyUI-Impact-Pack/blob/940ad41b1508cf9156bea8ee0ef7c87a851cc24f/js/impact-pack.js#L240

**Reply by devingfx** [reply]
Hi!

I managed to have a wildcard RETURN_TYPE and to cast the type of the returned value according to the one COMBO input in backend, great!

On the frontend side, I managed to set the output's type and name according to the COMBO input in `onConnectionsChange` and also in the COMBO's `callback` for the thing to be dynamic, but it's still buggy:

- if the output that was of a type (let's say STRING) is connected to an input of same type, and the COMBO value (therefore the output type) changes, the link should be invalidated. Is there a way to mark the link (or graph?) dirty to make the link as invalid? Or I need to manually delete it? (In Blender, great implementation of Nodes & Grapth, if the slots changes dynamicly and the type don't work anymore, the link is colored in red with a tooltip, and if a slot is deleted dynamicly, then the associated links disappear, but if the node gets its slot back the link come back again...)
- the output name is the same as the current type, but if the graph is serialized, and then loaded again (aka F5) the output's name is set to the default again whatever it was cahnged to dynamicly. Could you point me what event/callback I need to hijack for the load of the node widgets to execute my dynamic output's type/name setter, please?

**Reply by christian-byrne** [reply]
> Could you point me what event/callback I need to hijack for the load of the node widgets to execute my dynamic output's type/name setter, please?

See the `onGraphConfigured` and `onAfterGraphConfigured` callbacks which you can hijack: https://github.com/Comfy-Org/ComfyUI_frontend/blob/56f3842045758de6c6029915525df013d2cf6f4a/src/scripts/app.ts#L1705-L1728

[Example](https://github.com/Comfy-Org/ComfyUI_frontend/blob/56f3842045758de6c6029915525df013d2cf6f4a/src/extensions/core/uploadAudio.ts#L171-L178)

**Reply by devingfx** [reply]
@christian-byrne What is the difference between overriding `LGraphNode.prototype.onConfigure` and `ComfyNode<nodeType>.prototype.onGraphConfigure/onAfterGraphConfigured` ?

**Reply by christian-byrne** [reply]
> @christian-byrne What is the difference between overriding `LGraphNode.prototype.onConfigure` and `ComfyNode<nodeType>.prototype.onGraphConfigure/onAfterGraphConfigured` ?


To my knowledge, `ComfyNode` class is not exposed for you to overrwrite properties on. Assuming by `ComfyNode<nodeType>` you meant an instance of a ComfyNode, then you will need to access via `ComfyNode<nodeType>.constructor.prototype`. The target for the modifications will depend on desired scope of changes: 

- Apply to a specific instance of a particular node: `ComfyNode<nodeType>.onGraphConfigured/onAfterGraphConfigured`
- Apply to all current or future instances of a particular node: `ComfyNode<nodeType>.constructor.prototype.onGraphConfigured/onAfterGraphConfigured`
- Apply to all instances of all nodes: `LGraphNode.prototype.onAfterGraphConfigured`

**Reply by devingfx** [reply]
by `ComfyNode<nodeType>` I meant a subclass of `ComfyNode` not a instance... BTW all methods are accessible because of inheritance, I wrote that this way to somehow point out where each methods was defined...

The question was more: "Why not override `onConfigure` on a specific ComfyNode subclass (and safely calling the old/overriden method to let the inheritance chain apply)? Is it safe/dangerous?"

Meanwhile I figured out and opened an [issue](https://github.com/Comfy-Org/ComfyUI_frontend/issues/784), that is, on my undestanding (and debug session step by step), a core extension is overriding `onGraphConfigured` too hardly (not calling the overriden method) and extensions overriding this cannot be called. Because the order of extension loading is not garantied due to async operation... So in my case, my extension did override the method, then `widgetInput` is loaded after and completely erase my ext's overrding...

**Reply by christian-byrne** [reply]
Thanks, I understand now. `ComfyNode<nodeType>` to denote a subclass of `ComfyNode` is from what type system?

Can you give an example of where `ComfyNode` is subclassed?

**Reply by devingfx** [reply]
> Can you give an example of where `ComfyNode` is subclassed?

To speak extremly precisely: There is no `ComfyNode` class that can be subclassed. The real class that is subclassed is `LGraphNode`. Each backend's nodedefs is used to create a subclass of `LGraphNode` that are all named `ComfyNode` (the `Function.prototype.name`) but are in fact a particular class for a given nodedef... What I called earlier `ComfyNode<nodeType>` instead of saying `AParticualrExtensionBackendNodeComfyNode` class for a particular python class' nodedef...

This is done here:

https://github.com/Comfy-Org/ComfyUI_frontend/blob/56f3842045758de6c6029915525df013d2cf6f4a/src/scripts/app.ts#L2016

(each `const node` is a class (Function) that create a particular `LGraphNode` according to the corresponding loaded nodedef, and not an instance of node. Instances of nodes are created using this registered class by the UI when you add a node...)

> ... is from what type system?
This type notation is same syntax as TypeScript's type annotation, but, once again, this is a way speaking, it is not a proper ts type...

**Reply by christian-byrne** [reply]
Thanks. It's a good explanation.
